### PR TITLE
Orchestrator (PREVIEW) docs and samples

### PR DESCRIPTION
**Pull request by vishwacsena**
Add Orchestrator [preview] docs and samples

All content goes under \experimental\orchestrator.

### CODE REVIEW

File: experimental/orchestrator/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        modelPath: new StringExpression(process.env.ModelPath),
+        snapshotPath: new StringExpression(process.env.RootDialogSnapshotPath),
+    });
+})
+```
+
+## Additional reading
+- [Tech overview][18]
+- [API refrence][14]
```
Comment (BruceHaley) [inline experimental/orchestrator/README.md:121]: Typo: "reference"

File: experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# Declarative bot
+
+This sample demonstrates using [Adaptive dialog][1],  [Language Generation][2] features with Orchestraor to demonstrate an end-to-end Alarm bot in action.
```
Comment (BruceHaley) [inline experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/README.md:3]: Typo: "Orchestrator"

File: experimental/orchestrator/docs/Overview.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
+# Technical Overview
+
+The Orchestrator is a replacement of the [Bot Framework Dispatcher][1] used in chat bots since 2018. It makes the state of the art natural language understanding methods available to bot developers while at the same time making the process of language modeling quick, not requiring the expertise in [Deep Neural Networks (DNN), Transformers][5], or [Natural Language Processing (NLP)][6]. This work is co-authored with the industry experts in the field and include some of the top methods used in the [General Language Understanding Evaluation (GLUE)][7] leaderboard. Orchestrator will continue evolve and adopt the latest advancements in science and in the industry.
```
Comment (BruceHaley) [inline experimental/orchestrator/docs/Overview.md:3]: Typos: "includes", "...will continue to evolve..."

File: experimental/orchestrator/README.md (hunk 1 above)
Comment (vishwacsena) [reply experimental/orchestrator/README.md:121]: Fixed.

File: experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/README.md (hunk 2 above)
Comment (vishwacsena) [reply experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/README.md:3]: Fixed.

File: experimental/orchestrator/docs/Overview.md (hunk 3 above)
Comment (vishwacsena) [reply experimental/orchestrator/docs/Overview.md:3]: Fixed.

File: experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+
+namespace Microsoft.BotBuilderSamples
+{
+    public static class DialogExtensions
+    {
+        public static async Task Run(this Dialog dialog, ITurnContext turnContext, IStatePropertyAccessor<DialogState> accessor, CancellationToken cancellationToken = default(CancellationToken))
+        {
```
Comment (stevengum) [inline experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs:14]: @vishwacsena do we need this Run() as opposed to [DialogExtensions.RunAsync()](https://github.com/microsoft/botbuilder-dotnet/blob/54cc02627e4d0ffbe8ae68ea3f534bdc7aae4386/libraries/Microsoft.Bot.Builder.Dialogs/DialogExtensions.cs#L31) in the shipped packages?

File: experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/Dialogs/ViewToDoDialog/ViewToDoDialog.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            };
+
+            // Add named dialogs to the DialogSet. These names are saved in the dialog state.
+            AddDialog(viewToDoDialog);
+
+            // The initial child Dialog to run.
+            InitialDialogId = nameof(AdaptiveDialog);
+        }
+
+        private static Recognizer CreateLuisRecognizer(IConfiguration configuration)
```
Comment (stevengum) [inline experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/Dialogs/ViewToDoDialog/ViewToDoDialog.cs:86]:
```suggestion
        private static Recognizer CreateOrchestratorRecognizer(IConfiguration configuration)
```

File: experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        public DeclarativeBotAdapter(ICredentialProvider credentialProvider, IConfiguration configuration, ILogger<BotFrameworkHttpAdapter> logger, IStorage storage, UserState userState, ConversationState conversationState, ResourceExplorer resourceExplorer)
+            : base(credentialProvider)
+        {
+            this.Use(new RegisterClassMiddleware<IConfiguration>(configuration));
+            this.UseStorage(storage);
+            this.UseBotState(userState);
+            this.UseBotState(conversationState);
+
+            this.OnTurnError = async (turnContext, exception) =>
```
Comment (stevengum) [inline experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs:24]: Are `this.` part of the samples' StyleCop configuration?

File: experimental/orchestrator/docs/Overview.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+#### Local, fast library not a remote service
+
+The Orchestrator core is written in C++ and is available as a library in C#, Node.js and soon Python and Java. The library can be used directly by the bot code (a preferred approach) or can be hosted out-of-proc or on a remote server. Loading the English pretrained language model released for the initial preview takes about 2 sec with the memory footprint of a little over 200MB. Classification of a new example with this initial model takes about 10 milliseconds (depending on the text length). These numbers are for illustration only to give a sense of performance. As we improve the models or include additional languages these numbers will likely change.
```
Comment (stevengum) [inline experimental/orchestrator/docs/Overview.md:23]: `The library can be used directly by the bot code (a preferred approach) or can be hosted out-of-proc or on a remote server.`

Why is it preferred to be locally consumed by the bot?

File: experimental/orchestrator/docs/Overview.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+Orchestrator enables composability of bots allowing reuse of skills or the entire bots contributed by the community in an easy way without requiring time consuming retraining of the language models. It is our goal to support the community and continue responding to the provided feedback.
+
+## Design Objectives and Highlights
+
+Thanks to the community feedback we compiled a list of objectives and requirements which are addressed in the initial release of the Orchestrator. The [Roadmap](###roadmap) section describes the additional work planned in the upcoming releases. 
+
+#### No [ML][12] or [NLP][6] expertise required
```
Comment (stevengum) [inline experimental/orchestrator/docs/Overview.md:11]: Personal style suggestion, I would use `###` as opposed to `####` as three stand out better, and better indicate the importance of each section.

E.g.

### No [ML][12] or [NLP][6] expertise required

vs 

#### No [ML][12] or [NLP][6] expertise required

[12]: https://github.com/microsoft/BotBuilder-Samples/pull/2674
[6]: https://github.com/microsoft/BotBuilder-Samples/pull/2674

File: experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+// Import required bot services.
+// See https://aka.ms/bot-services to learn more about the different parts of a bot.
+const { BotFrameworkAdapter } = require('botbuilder');
+
+const { DispatchBot } = require('./bots/dispatchBot');
+
+// Note: Ensure you have a .env file and include all necessary credentials to access services like LUIS and QnAMaker.
+const ENV_FILE = path.join(__dirname, '.env');
+require('dotenv').config({ path: ENV_FILE });
```
Comment (stevengum) [inline experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js:18]: This load order seems problematic; we should ensure the environment variables are loaded before we import any local files.

Comment (vishwacsena) [reply experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js:18]: Not sure I follow. Can you please clarify? This is the exact same pattern we use in luis/ qnamaker samples as well as the dispatch sample we have for js_nodejs.

https://github.com/microsoft/BotBuilder-Samples/blob/6ba153cab7426e4061e3283037ef0091d481a43e/samples/javascript_nodejs/13.core-bot/index.js#L10

File: experimental/orchestrator/docs/Overview.md (hunk 8 above)
Comment (vishwacsena) [reply experimental/orchestrator/docs/Overview.md:11]: Updated.

File: experimental/orchestrator/docs/Overview.md (hunk 7 above)
Comment (vishwacsena) [reply experimental/orchestrator/docs/Overview.md:23]: Clarified.

File: experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/Dialogs/ViewToDoDialog/ViewToDoDialog.cs (hunk 5 above)
Comment (vishwacsena) [reply experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/Dialogs/ViewToDoDialog/ViewToDoDialog.cs:86]: Fixed.

File: experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs (hunk 4 above)
Comment (vishwacsena) [reply experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs:14]: Not sure. But would like to keep it this way for preview 1. This is consistent with the current adaptive samples.

File: experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs (hunk 6 above)
Comment (vishwacsena) [reply experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs:24]: I started with the existing adaptive samples. I'm open to changing/ updating to be consistent but we need to do it across the board. Let's file a ticket to track any code style related fixes required to this + adaptive samples?

File: experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs (hunk 4 above)
Comment (stevengum) [reply experimental/orchestrator/csharp_dotnetcore/02.todo-bot-with-adaptive-dialog/DialogExtensions.cs:14]: > Not sure. But would like to keep it this way for preview 1. This is consistent with the current adaptive samples.

Sounds good to me

File: experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs (hunk 6 above)
Comment (stevengum) [reply experimental/orchestrator/csharp_dotnetcore/03.declarative-bot/DeclarativeBotAdapter.cs:24]: Filed https://github.com/microsoft/BotBuilder-Samples/issues/2681 for investigating style in .NET samples

File: experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js (hunk 9 above)
Comment (stevengum) [reply experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js:18]: Of course, it's the order of imports and loading of environment variables that could potentially cause problems. In JS Core bot, the `.env` file is loaded before any business logic is imported into the `index.js`:

https://github.com/microsoft/BotBuilder-Samples/blob/6ba153cab7426e4061e3283037ef0091d481a43e/samples/javascript_nodejs/13.core-bot/index.js#L10-L26

In the dispatch sample, `.env` values are loaded after the business logic is imported, so when you access `process.env` in the business logic matters:

https://github.com/microsoft/BotBuilder-Samples/blob/7a29b732c94e7ec74a4b9bc68447aafbea791f01/experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js#L14-L18

Christopher filed an issue for this: https://github.com/microsoft/BotBuilder-Samples/issues/2374, and @Zerryth addressed the issue in the generators and corresponding samples.

Comment (vishwacsena) [reply experimental/orchestrator/javascript_nodejs/01.dispatch-bot/index.js]: Thanks @stevengum .. I've updated the samples to have env load as early as practicable.

**Pull request merged**